### ISSUE TITLE
Misleading HKT answer in FAQ

### ISSUE THREAD

**Issue by withoutboats** [rust-lang/rust-www#401]
The answer in the FAQ about higher-kinded polymorphism contains some misleading information.

> Support for higher-kinded types means these “incomplete” types may be used anywhere “complete” types can be used, including as generics for functions.

This is a very standard explanation of higher kinded polymorphism, but it is also very incorrect (and why I strongly prefer avoiding the term "higher-kinded type" in general). Higher-kinded polymorphism doesn't allow using a type constructor anywhere a type can be used - the most obvious and enormous case being that type constructors cannot be used __as types__.

>  A type with one parameter, like Vec<T> is of kind * -> *, meaning that Vec<T> takes in a complete type like i32 and returns a complete type Vec<i32>.

The use of the syntax `Vec<T>` to refer to the type constructor seems confusing to me. We wouldn't say `not(x)` is a "a function of the type bool -> bool", we would say `not` is a function of the type bool -> bool. This gives readers a false impression about the relationship between higher-kindedness and type parameters, and can lead them down a path of confusion.

> A type with three parameters, like HashMap<K, V, S> is of kind * -> * -> * -> *, and takes in three complete types (like i32, String, and RandomState) to produce a new complete type HashMap<i32, String, RandomState>.

Other than the same syntax issue as my previous comment, this uses currying, which Rust is very unlikely to adopt in its kind semantics. It should at least be written `(*, *, *) -> *`

I also think the `*` and `Lt` syntax is less preferable than just writing out `type` and `lifetime` in these kinds.

The remainder is accurate but I'm not sure its informative - it doesn't really explain what "collections need higher kinded polymorphism" actually looks like, and "monad" appears ominously without any further explanation.

I'm not sure this subject is good material for an FAQ. Higher kinded polymorphism is a rather advanced subject of type theory, not suited to an explanation in the length of an FAQ answer, Moreover, this answer is trying to explain why a feature Rust does not have, and is not actively working toward having, is a good thing Rust should have - that's a sort of advocacy I don't think is appropriate for this document.

**Comment by AndrewBrinker** [comment]
Yeah, this answer was one of the most heavily-discussed and heavily-questioned answers in the FAQ. It was only included because it _is_ a frequently asked question about Rust. If you'd like to take a crack at a better answer, that would be great. I'm also happy to work with you to put together a better answer.

**Comment by withoutboats** [comment]
@AndrewBrinker I'd be happy to work on an answer with you :-).

In relation to the [RFC](https://github.com/rust-lang/rfcs/pull/1598) I authored, I've been working on a blog series explaining higher kinded polymorphism in detail - it grew into a series of 5 posts, 1 explaining kindedness, 3 explaining the different kinds of higher kinded polymorphism Rust could support, and a final one expressing my own opinion about what would be best for Rust to do. I want to break down exactly what this question is trying to answer, because I'm worried there is just too much to explain to answer this question in an FAQ.

Here are the elements of this question right now:

* What is higher kinded polymorphism?
    * What is a kind? What are terms of a higher kind?
    * What does it mean to be polymorphic over terms of higher kind?
* What is the status of higher kinded polymorphism in Rust?

Basically, I think the second question is easily answered in the FAQ, but I don't think the first fits in this format. Maybe its possible to word the second in a way that is still informative if you don't properly grasp what higher kinded polymorphism means?

**Comment by AndrewBrinker** [comment]
Okay, I finally have some time to work on this! I'll include this in the swath of new answers and answer-improvements I'm working on right now. @withoutboats, I'll let you know as soon as I have a draft.

**Comment by withoutboats** [comment]
@AndrewBrinker Cool. I think we also have a clearer picture of what direction Rust is going in for this feature (short answer: mostly along the lines of RFC 1598, linked above) than we did a few months ago. :-)

**Comment by AndrewBrinker** [comment]
Great! Should make writing up a solid answer a bit easier.
